vault/identity_store_oidc.go

```diff
@@ -0,0 +1,1349 @@
+package vault
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"gopkg.in/square/go-jose.v2/jwt"
+
+	"github.com/hashicorp/go-hclog"
+
+	"github.com/hashicorp/vault/sdk/helper/strutil"
+
+	"github.com/hashicorp/errwrap"
+	"github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/vault/helper/identity"
+	"github.com/hashicorp/vault/sdk/framework"
+	"github.com/hashicorp/vault/sdk/logical"
+	"gopkg.in/square/go-jose.v2"
+)
+
+type oidcConfig struct {
+	Issuer string `json:"issuer"`
+
+	// effectiveIssuer is a calculated field and will be either Issuer (if
+	// that's set) or the Vault instance's api_addr.
+	effectiveIssuer string
+}
+
+type expireableKey struct {
+	KeyID    string    `json:"key_id"`
+	ExpireAt time.Time `json:"expire_at"`
+}
+
+type namedKey struct {
+	Name            string           `json:"name"`
+	Algorithm       string           `json:"signing_algorithm"`
+	VerificationTTL time.Duration    `json:"verification_ttl"`
+	RotationPeriod  time.Duration    `json:"rotation_period"`
+	KeyRing         []*expireableKey `json:"key_ring"`
+	SigningKey      *jose.JSONWebKey `json:"signing_key"`
+	NextRotation    time.Time        `json:"next_rotation"`
+}
+
+type role struct {
+	Name     string        `json:"name"` // TODO: do we need/want this?
+	TokenTTL time.Duration `json:"token_ttl"`
+	Key      string        `json:"key"`
+	Template string        `json:"template"`
+	RoleID   string        `json:"role_id"`
```

Comment:
It's in a role struct so this could just be ID.

Reply:
Changing it to ClientID, per later comment.